File: chaincode/src/model/yellowPages.ts
/*
 * SPDX-License-Identifier: Apache-2.0
 */
import * as Yup from 'yup';

export class YellowPages {

    public static readonly schema = Yup.object().shape({
        docType: Yup.string().notRequired(),
        originAutomationRegisteredResourceMrid: Yup.string().required(
            'originAutomationRegisteredResourceMrid is a compulsory string.'),
        registeredResourceMrid: Yup.string().required(
            'registeredResourceMrid is a compulsory string.'),
        systemOperatorMarketParticipantMrid: Yup.string().required(
            'systemOperatorMarketParticipantMrid is a compulsory string.'),
        yellowPageMrid: Yup.string().required(
            'yellowPageMrid is a compulsory string.'),
        });

    public docType?: string;
    public yellowPageMrid: string;
    public originAutomationRegisteredResourceMrid: string; // FK1
    public registeredResourceMrid: string; // FK2
    public systemOperatorMarketParticipantMrid: string; // FK3
}

File: backend/service/src/main/java/com/star/models/yellowpages/YellowPages.java
package com.star.models.yellowpages;

import com.fasterxml.jackson.annotation.JsonIgnore;
import com.star.models.imports.ImportCSV;
import lombok.AllArgsConstructor;
import lombok.Builder;
import lombok.Data;
import lombok.NoArgsConstructor;
import org.apache.commons.csv.CSVRecord;

import javax.validation.constraints.NotBlank;
import java.util.Arrays;
import java.util.List;

import static java.util.UUID.randomUUID;

/**
 * Copyright (c) 2022, Enedis (https://www.enedis.fr), RTE (http://www.rte-france.com)
 * SPDX-License-Identifier: Apache-2.0
 */
@Builder
@Data
@NoArgsConstructor
@AllArgsConstructor
public class YellowPages implements ImportCSV {

    @JsonIgnore
    private final List<String> headers = Arrays.asList(
            "originAutomationRegisteredResourceMrid",
            "registeredResourceMrid",
            "systemOperatorMarketParticipantMrid"
    );
    @JsonIgnore
    private String docType;
    @NotBlank(message = "Le champ yellowPageMrid est obligatoire")
    private String yellowPageMrid;
    @NotBlank(message = "Le champ originAutomationRegisteredResourceMrid est obligatoire")
    private String originAutomationRegisteredResourceMrid;
    @NotBlank(message = "Le champ registeredResourceMrid est obligatoire")
    private String registeredResourceMrid;
    @NotBlank(message = "Le champ systemOperatorMarketParticipantMrid est obligatoire")
    private String systemOperatorMarketParticipantMrid;

    @Override
    public List<String> getHeaders() {
        return headers;
    }

    @Override
    public void defineData(CSVRecord csvRecord) {
        if (csvRecord == null) {
            throw new IllegalArgumentException("Can not instantiate data from a null csvRecord");
        }
        this.yellowPageMrid = randomUUID().toString();
        this.originAutomationRegisteredResourceMrid = csvRecord.get(headers.get(0));
        this.registeredResourceMrid = csvRecord.get(headers.get(1));
        this.systemOperatorMarketParticipantMrid = csvRecord.get(headers.get(2));
    }

}